base/compiler/abstractinterpretation.jl

```diff
@@ -72,6 +72,11 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
     any_const_result = false
     const_results = Union{InferenceResult,Nothing,ConstResult}[]
     multiple_matches = napplicable > 1
+    if matches.overlayed
```

Comment:
How do we know that a non-overlayed function that had effects inferred as concrete-eval eligible wasn't supposed to eventually call an overlayed function. I think we'd need to track that as an effect in order to be able to do this.

Reply:
Shouldn't that simply be established by inference consistency? If you looked in the cache and it said the method has certain properties, but you have not actually done the work to compute that correctly, you are already in very deep trouble.

Reply:
Oh, you mean it was correctly inferred (with using a overlayed function), so we need to track whether the normal execution (at compile time here) will give exactly the same answer as running the overlayed transform (at runtime)?

Reply:
yeah, if I understand correctly, this PR needs to correctly handle this test case (, which doesn't pass currently):
```julia
# disable partial pure/concrete evaluation when tainted by any overlayed call
strangesin(x) = sin(x)
@overlay OverlayedMT strangesin(x::Float64) = iszero(x) ? nothing : cos(x)
Base.@assume_effects :total totalcall(f, args...) = f(args...)
@test Base.return_types(; interp=MTOverlayInterp()) do
    if totalcall(strangesin, 1.0) == cos(1.0)
        return nothing
    else
        return missing
    end
end |> only === Nothing
```
This specific test case additionally requires #44561 though.

Reply:
That's one example, yes.

Reply:
Addressed.